Tidy up `pytket.circuit` API docs

CalMacCQ opened the pull request:
# Description

First pass of cleaning up the `pytket.circuit` documentation, still plenty more to do.

* Rather than autopopulating the circuit class docs with all methods I've used `.. automethod::` to add each `Circuit` method individually. This means we can move the methods we want people to use to the top of the docs (i.e. Circuit.add_gate) and move others to the bottom (i.e `Circuit.add_add_pauliexpcommutingsetbox` and `Circuit.CX`)

If anyone knows how to get control of the ordering without using `automethod` hundreds of times then PLEASE let me know.

* added some notes saying that `add_gate` is sufficent to add append any box to a circuit rather than using `Circuit.add_x_box`. Also saying that using the convenices methods like `Circuit.CX(0, 1)` are equiavlent to using `Circuit.add_gate(OpType.CX, [0, 1])` .

* Don't clutter the `pytket.circuit` page by expanding the `logic_exp` classes


before

![Screenshot 2023-12-15 at 16 39 37](https://github.com/CQCL/tket/assets/93673602/de645b8f-9bd8-4517-89fd-207696476177)

after

![Screenshot 2023-12-15 at 16 44 16](https://github.com/CQCL/tket/assets/93673602/4ccaa9d2-ed06-4a72-9a53-84201e89b512)



* Populated the right sidebar to show the contents of each pytket module (see image). This is done globally in the docs by setting `"show_toc_level": 2` in `html_theme_options`.

before 

![Screenshot 2023-12-15 at 16 31 21](https://github.com/CQCL/tket/assets/93673602/e2b5ab5e-0692-443a-bc3d-8cf3ff5264fd)

after

![Screenshot 2023-12-15 at 16 30 40](https://github.com/CQCL/tket/assets/93673602/5b0bea37-f31a-4469-88dd-f8af8972c1d6)


# Checklist
- [x] I have performed a self-review of my code.
- [x] I have commented hard-to-understand parts of my code.
- [ ] I have made corresponding changes to the public API documentation.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have updated the changelog with any user-facing changes.

CalMacCQ commented on pytket/docs/circuit.rst:
```diff
@@ -14,104 +14,104 @@ pytket.circuit
 
 .. automodule:: pytket._tket.circuit
     :members: fresh_symbol
-.. autoclass:: pytket._tket.circuit.Op
```
I think all of this _tket stuff is replaced in the docs build anyway. I don't think theres any good reason for it to be in the rst source

CalMacCQ commented on pytket/docs/conf.py:
```diff
@@ -132,78 +134,6 @@
-    # The paper size ('letterpaper' or 'a4paper').
-    #
-    # 'papersize': 'letterpaper',
-    # The font size ('10pt', '11pt' or '12pt').
```
Don't think any of this does anything we care about.

CalMacCQ commented on pytket/docs/circuit.rst:
```diff
@@ -1,8 +1,8 @@
-   :caption: Classes:
-   :maxdepth: 2
+   :caption: Submodules:
+   :maxdepth: 1
```
seeting the depth to 1 ensures that [this page](https://tket.quantinuum.com/api-docs/circuit.html) isn't cluttered with all of the classes.

All of the contents can now be found in the right sidebar of the docs instead

CalMacCQ replied:
Changed it back to `Classes` for now

CalMacCQ commented on pytket/docs/conf.py:
```diff
@@ -99,6 +100,7 @@
 #
 
 html_theme_options = {
+    "show_toc_level": 2,
```
This config setting populates the right sidebar for all of the docs.

CalMacCQ commented on pytket/docs/conf.py:
```diff
@@ -55,6 +55,7 @@
 extensions = [
     "sphinx.ext.autodoc",
     "sphinx.ext.autosummary",
+    # "sphinx_autodoc_typehints",
```
I added this but it wasn't working properly so I've deactivated it for now.

ss2165 commented:
>  without using automethod hundreds of times 

we could use `bysource` ordering and just reorder the methods in the source?

CalMacCQ commented on pytket/docs/circuit_class.rst:
```diff
@@ -20,7 +20,322 @@ following keyword arguments are supported:
-   :members:
+.. currentmodule:: pytket.circuit.Circuit
+.. autoclass:: pytket.circuit.Circuit
+   :special-members: __init__, __eq__, __iter__, __mul__, __repr__, __rshift__, __str__, __getstate__, __setstate__, __hash__
```
Good point about the `__init__` method. Will try to fix that.

I'm not sure what you mean by 
```
It should also mainly be examples to show what each way of initialising means
```

Could you reword please?

CalMacCQ replied:
`__init__`, `__eq__` etc all now appear at the top. I've removed the `:special-members:` and used `:automethod:` following this answer -> https://stackoverflow.com/questions/42214656/sphinx-naming-special-members

ss2165 replied:
![image](https://github.com/CQCL/tket/assets/12997250/6f48cfd8-cebb-4d0d-ba8b-5215bddec9ca)

This is quite hard to understand, and constructing a circuit is one of the most important API entry points.
It would be better to go through a series of examples, something like:


```python
# initialise an empty circuit
c = Circuit()
# Initialise a circuit with 3 qubit
c = Circuit(3)
# initialise a circuit with 3 qubits and 3 bits
c = Circuit(3, 3)
# Optionally give a circuit a name
c = Circuit(3, name="mycircuit")

CalMacCQ replied:
Added some snippets on initialisation

ss2165 replied:
where? I can't find it

CalMacCQ replied:
Reverted the change as there was a weird issue I was having with the `pytket/binders/circuit/Circuit/main.cpp` file that was causing the CI checks to fail. Didn't figure out why.

Maybe I can add these in in a separate PR.

CalMacCQ commented on pytket/docs/circuit_class.rst:
```diff
@@ -20,7 +20,322 @@ following keyword arguments are supported:
+
+   .. automethod:: flatten_registers
+
+   .. autoproperty:: n_qubits
+
+   .. autoproperty:: n_bits
+
+   .. autoproperty:: phase
+
+   .. autoproperty:: qubits
+
+   .. autoproperty:: bits
+
```
Good point, done

CalMacCQ commented:
> > without using automethod hundreds of times
> 
> we could use `bysource` ordering for `autoclass` and just reorder the methods in the source?

Using `bysource` this way seems to require pure python modules unfortunately -> https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#confval-autodoc_member_order

CalMacCQ commented on pytket/binders/circuit/Circuit/main.cpp:
```diff
@@ -119,8 +121,8 @@ void def_circuit(py::class_<Circuit, std::shared_ptr<Circuit>> &pyCircuit) {
       .def(
           py::init<unsigned, unsigned, std::optional<std::string>>(),
-          "Constructs a circuit with a given number of quantum and "
-          "classical bits\n\n:param n_qubits: The number of qubits in "
```
Most of these changes are autoformatting? Not sure why the diff is so big with only a minor edit.

cqc-alec reviewed (commented):
Having made a start at cross-checking these circuit docs with the existing ones, I think we need a script to do it for us.

I found the following missing `Circuit` methods, before I stopped looking: `CU1`, `CU3`, `Reset`, `TK2`, `ZZMax`, `add_c_not`.

Could you perhaps write a script that iterates over all `Circuit` methods (use `Circuit.__dict__.keys()`) and lists any that do not occur in `circuit_class.rst`?

ss2165 commented:
> Could you perhaps write a script that iterates over all Circuit methods (use Circuit.__dict__.keys()) and lists any that do not occur in circuit_class.rst?


```python
# print all members of Circuit that don't start with underscore that don't occur in the docs
import inspect
import re

from pytket._tket.circuit import Circuit

# paste contents of https://github.com/CQCL/tket/blob/7b0df5cf68503944cd34e593fb094ba5aa4357f4/pytket/docs/circuit_class.rst
rst = ...
in_methods = {p[0] for p in inspect.getmembers(Circuit)}
pattern = re.compile(r"(?:\bautomethod\b|\bautoproperty\b):: (\w+)\n")
in_docs = set(pattern.findall(rst))
missing = in_methods - in_docs
print([s for s in missing if not s.startswith("_")])

```
Output:
```python
['CU1', 'discarded_qubits', 'bit_readout', 'Reset', 'valid_connectivity', 'CU3', 'add_multiplexed_tensored_u2', 'symbol_substitution', 'add_conditional_barrier', 'add_c_not', 'qubit_to_bit_map', 'ZZMax', 'created_qubits', 'depth_by_type', 'is_symbolic', 'add_phasepolybox', 'name', 'substitute_named', 'TK2']
```

CalMacCQ commented:
Now added the missing methods found by the script. Not sure if its worth using the scipt as part of the CI checks or whether we should just do a one-off check.

It seems to show `depth_by_type` as still missing but its in there.

cqc-alec commented:
> I've added the missing methods found by the script.
> 
> Not sure if its worth using the script as part of the CI checks or whether we should just do a one-off check.
> 
> The script seems to show `depth_by_type` as still missing from `circuit_class.rst` but its in there.

That's very strange.

Can we check in the script to the repo, as it will be useful in future even if we don't run it on the CI?

CalMacCQ commented:
> > I've added the missing methods found by the script.
> > Not sure if its worth using the script as part of the CI checks or whether we should just do a one-off check.
> > The script seems to show `depth_by_type` as still missing from `circuit_class.rst` but its in there.
> 
> That's very strange.
> 
> Can we check in the script to the repo, as it will be useful in future even if we don't run it on the CI?

Done, added the script.

cqc-alec commented on pytket/docs/circuit_class.rst:
```diff
@@ -20,7 +20,375 @@ following keyword arguments are supported:
+
+   .. automethod:: depth_2q
+
+   .. automethod:: depth_by_type 
```
There's a space at the end of this line: that's why the script wasn't detecting it.

CalMacCQ replied:
thanks, fixed it

cqc-alec commented on pytket/docs/check_circuit_class_docs.py:
```diff
@@ -0,0 +1,16 @@
+
+from pytket._tket.circuit import Circuit
+
+# paste contents of https://github.com/CQCL/tket/blob/7b0df5cf68503944cd34e593fb094ba5aa4357f4/pytket/docs/circuit_class.rst
```
I don't think we want this comment?

CalMacCQ replied:
removed

CalMacCQ commented on pytket/docs/circuit_class.rst:
```diff
@@ -20,7 +20,375 @@ following keyword arguments are supported:
+   .. automethod:: to_latex_file
+
+
+   Convenience methods for appending circuit operations
```
Made it a heading

cqc-alec commented on pytket/docs/check_circuit_class_docs.py:
Can we put a couple of lines of commentary at the top of this file to explain what it does and why?

CalMacCQ replied:
Good idea, done.
